src/SSHDebugPS/CommandRunner.cs

```diff
@@ -9,304 +9,304 @@
-
-        public void WriteLine(string text)
+        protected void ReportException(Exception ex)
         {
-            if (_isClosed)
-                return;
-            _stdoutWriter.WriteLine(text);
-            _stdoutWriter.Flush();
+            ErrorOccured?.Invoke(this, new ErrorOccuredEventArgs(ex));
         }
 
-        private void OnProcessExited(object sender, EventArgs e)
+        protected virtual void OnProcessExited(object sender, EventArgs args)
         {
             lock (_lock)
             {
-                if (!_hasExited && _outputReadTask.IsCompleted)
+                // Make sure that all output has been written before exiting.
+                if (!_hasExited && _outputReadLoopTask.IsCompleted)
                 {
                     _hasExited = true;
-                    Closed?.Invoke(this, _localProcess.ExitCode);
+                    Closed?.Invoke(this, _process.ExitCode);
                 }
             }
         }
-    }
-
-    /// <summary>
-    /// Launches a local command that sends output when a newline is received.
-    /// </summary>
-    internal class LocalBufferedCommandRunner : ICommandRunner
-    {
-        private System.Diagnostics.Process _localProcess;
-        private StreamWriter _stdoutWriter;
-        private ProcessStartInfo processStartInfo;
-
-        private bool _isClosed = false;
 
-        public LocalBufferedCommandRunner(string command, string arguments)
+        protected virtual void OnErrorReceived(string error)
         {
-            processStartInfo = new ProcessStartInfo(command, arguments);
-            processStartInfo.RedirectStandardError = true;
-            processStartInfo.RedirectStandardInput = true;
-            processStartInfo.RedirectStandardOutput = true;
-            processStartInfo.UseShellExecute = false;
-            processStartInfo.CreateNoWindow = true;
-
-            _localProcess = System.Diagnostics.Process.Start(processStartInfo);
-            _stdoutWriter = new StreamWriter(_localProcess.StandardInput.BaseStream, new UTF8Encoding(encoderShouldEmitUTF8Identifier: false), 4096, leaveOpen: false);
-            _localProcess.OutputDataReceived += OnProcessOutput;
-            _localProcess.ErrorDataReceived += OnErrorOutput;
-            _localProcess.Exited += OnProcessExited;
-
-            _localProcess.BeginOutputReadLine();
-            _localProcess.BeginErrorReadLine();
-            _localProcess.EnableRaisingEvents = true;
+            ErrorOccured?.Invoke(this, new ErrorOccuredEventArgs(error));
         }
 
-        public event EventHandler<string> OutputReceived;
-        public event EventHandler<int> Closed;
-        public event EventHandler<ErrorOccuredEventArgs> ErrorOccured;
+        protected virtual void OnOutputReceived(string line)
+        {
+            OutputReceived?.Invoke(this, line);
+        }
 
-        public void Dispose()
+        protected virtual void CreateProcessStartInfo(string command, string commandArgs)
```

Comment:
I don't think this method needs to be virtual, but if you want it to be, it needs to return the ProcessStartInfo instead of setting the private field directly, or else an implementation that overrides this method will blow up when you call Start, since the private field isn't set.

Reply:
I'll remove the virtual

---
In reply to: [297443578](https://github.com/microsoft/MIEngine/pull/884#discussion_r297443578) [](ancestors = 297443578)